src/shared_utilities/FileReader.jl

```diff
@@ -0,0 +1,371 @@
+"""
+    FileReader
+
+This module coordinates reading of input data from NetCDF files,
+and allows us to read in data containing a time component.
+It also includes regridding and temporal interpolations
+of this data.
+
+This is based on ClimaCoupler.jl's BCReader and TimeManager modules.
+"""
+module FileReader
+
+using ClimaComms
+using ClimaCore: Fields
+using Dates
+using JLD2
+import CFTime
+
+using ClimaLSM.Regridder
+
+export to_datetime,
+    datetime_to_strdate,
+    FileInfo,
+    file_info_init,
+    float_type_fi,
+    read_data_fields!,
+    next_date_in_file,
+    interpolate_data
+
+
+"""
+    to_datetime(date)
+
+Convert a DateTime-like object (e.g. DateTimeNoLeap) to a DateTime,
+using CFTime.jl. We need this since the CESM2 albedo file contains
+DateTimeNoLeap objects for dates, which can't be used for math with DateTimes.
+
+Note that this conversion may fail if the date to convert doesn't
+exist in the DateTime calendar.
+
+# Arguments
+- `date`: object to be converted to DateTime
+"""
+to_datetime(date) = CFTime.reinterpret(DateTime, date)
+
+"""
+    datetime_to_strdate(datetime::DateTime)
+
+Convert from Date to String ("YYYYMMDD") format.
+
+# Arguments
+- `datetime`: [DateTime] object to be converted to string
+"""
+datetime_to_strdate(datetime::DateTime) =
+    string(lpad(Dates.year(datetime), 4, "0")) *
+    string(string(lpad(Dates.month(datetime), 2, "0"))) *
+    string(lpad(Dates.day(datetime), 2, "0"))
+
+"""
+    FileInfo
+
+Stores information specific to each boundary condition from a file and each variable.
+
+# Inputs:
+- file_path::S                    # directory of the data file
+- comms::X                        # communication context used for MPI
+- varname::V                      # name of the variable
+- all_dates::D                    # vector of all dates contained in the original data file
+- data_fields::C                  # tuple of two fields at consecutive dates, that will be used for interpolation
+- surface_space::B                # the space we're reading data onto
+- date_idx::Vector{Int}           # time index of the data in the file
+- date_idx0::Vector{Int}          # `date_idx` of the file data that is closest to date0
+- segment_length::Vector{Int}     # length of each date segment (used in the daily interpolation)
+- regrid_dirpath::S               # filename root for regridded data
+- outfile_root::S                 # string root for saved output files for this variable
+"""
+struct FileInfo{FT <: Real, S, X, V, D, C, B, VI}
+    file_path::S
+    comms::X
+    varname::V
+    all_dates::D
+    data_fields::C
+    surface_space::B
+    date_idx::VI
+    date_idx0::VI
+    segment_length::VI
+    regrid_dirpath::S
+    outfile_root::S
+end
+
+FileInfo{FT}(args...) where {FT} = FileInfo{FT, typeof.(args[1:7])...}(args...)
+float_type_fi(::FileInfo{FT}) where {FT} = FT
```

Comment:
Just FYI: the float type is already stored in `Spaces.undertype(surface_space)`, so this type parameter is in a sense redundant

Reply:
Good point, I'll remove it